pkg/daemon/ceph/osd/remove.go

```diff
@@ -58,8 +59,25 @@ func RemoveOSDs(context *clusterd.Context, clusterInfo *client.ClusterInfo, osds
 		if status == upStatus {
 			logger.Infof("osd.%d is healthy. It cannot be removed unless it is 'down'", osdID)
 			continue
+		} else {
+			logger.Infof("osd.%d is marked 'DOWN'", osdID)
+		}
+
+		// Check we can remove the OSD in case jobs are fired in parallel
```

Comment:
It's not just about running multiple jobs in parallel, but also the OSD might not be safe to remove anyway, right?
```suggestion
		// Check if it is safe to remove the OSD
```

Reply:
I think we do a lot of our testing on systems where we are removing OSDs from disks that are fine. I'm curious how this could (or could not) negatively affect OSDs on disks that have failed. 

By my current understanding, if a disk has failed, Ceph should eventually still get into a place where it is okay to remove the failed OSD once data has been backfilled. I think that means that checking okay-to-stop is safe in the case of failed disks and that everything here is good. Does that all sound right?

Reply:
> I think we do a lot of our testing on systems where we are removing OSDs from disks that are fine. I'm curious how this could (or could not) negatively affect OSDs on disks that have failed.

From Ceph's perspective, I believe the `safe-to-stop` is just a check for whether at least n-1 replicas of the OSDs are still available even if this OSD goes down. Whether it's a hardware or software issue I would hope it doesn't make a difference in how we respond.

> By my current understanding, if a disk has failed, Ceph should eventually still get into a place where it is okay to remove the failed OSD once data has been backfilled. I think that means that checking okay-to-stop is safe in the case of failed disks and that everything here is good. Does that all sound right?

We really have two distinct commands in ceph to check (and my original comment was not clear about this separation either):
1. `ok-to-stop`: n-1 replicas are still available even if this OSD goes down
1. `ok-to-destroy`: The PGs from this OSD have all been backfilled to other OSDs and there is no risk in removing this OSD permanently. 

Stopping the OSD would be possible in any production cluster as long as other OSDs are healthy. It makes sense to wait for this condition by default. We should still have a flag that allows the OSD removal to be forced even if not safe to stop.

Destroying the OSD will only be considered ok if there are other OSDs to which the PGs can be backfilled, which is not possible in some clusters. By default I'm thinking we should not wait for this condition, but have a flag that waits for it to be ok to destroy.

Reply:
> Stopping the OSD would be possible in any production cluster as long as other OSDs are healthy. It makes sense to wait for this condition by default. We should still have a flag that allows the OSD removal to be forced even if not safe to stop.
> 
> Destroying the OSD will only be considered ok if there are other OSDs to which the PGs can be backfilled, which is not possible in some clusters. By default I'm thinking we should not wait for this condition, but have a flag that waits for it to be ok to destroy.

I think both of those points sound reasonable and good to me, and that info is helpful for my own understanding. 👍

Reply:
Actually, we need the `safe-to-destroy` call instead of ok-to-stop since, at this point when we run the job, the OSD is already stopped. Earlier in the script, if the OSD is not stopped we bailout. So we only need `safe-to-destroy` IMO.